src/Parallel/PhaseControl/VisitAndReturn.hpp

```diff
@@ -0,0 +1,193 @@
+
+  using phase_change_tags_and_combines =
+      tmpl::list<Tags::ReturnPhase<TargetPhase>,
+                 Tags::TemporaryPhaseRequested<TargetPhase>>;
+
+  template <typename LocalMetavariables>
+  using participating_components = typename LocalMetavariables::component_list;
+
+  template <typename... DecisionTags>
+  void initialize_phase_change_decision_data_impl(
+      const gsl::not_null<tuples::TaggedTuple<DecisionTags...>*>
+          phase_change_decision_data) const noexcept {
+    tuples::get<Tags::ReturnPhase<TargetPhase>>(*phase_change_decision_data) =
+        std::nullopt;
+    tuples::get<Tags::TemporaryPhaseRequested<TargetPhase>>(
+        *phase_change_decision_data) = false;
+  }
+
+  template <typename ParallelComponent, typename ArrayIndex>
+  void contribute_phase_data_impl(
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& array_index) const noexcept {
+    if constexpr (std::is_same_v<typename ParallelComponent::chare_type,
+                                 Parallel::Algorithms::Array>) {
+      Parallel::contribute_to_phase_change_reduction<ParallelComponent>(
+          tuples::TaggedTuple<Tags::TemporaryPhaseRequested<TargetPhase>>{true},
+          cache, array_index);
+    } else {
+      Parallel::contribute_to_phase_change_reduction<ParallelComponent>(
+          tuples::TaggedTuple<Tags::TemporaryPhaseRequested<TargetPhase>>{true},
+          cache);
+    }
+  }
+
+  template <typename... DecisionTags>
+  typename std::optional<
+      std::pair<typename Metavariables::Phase, ArbitrationStrategy>>
+  arbitrate_phase_control_impl(
+      const gsl::not_null<tuples::TaggedTuple<DecisionTags...>*>
+          phase_change_decision_data,
+      const typename Metavariables::Phase current_phase,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/) const noexcept {
+    auto& return_phase = tuples::get<Tags::ReturnPhase<TargetPhase>>(
+        *phase_change_decision_data);
+    if (return_phase.has_value()) {
+      const auto result = return_phase;
+      return_phase.reset();
+      return std::make_pair(result.value(),
+                            ArbitrationStrategy::PermitAdditionalJumps);
+    }
+    auto& temporary_phase_requested =
+        tuples::get<Tags::TemporaryPhaseRequested<TargetPhase>>(
+            *phase_change_decision_data);
+    if (temporary_phase_requested) {
+      return_phase = current_phase;
+      temporary_phase_requested = false;
+      return std::make_pair(TargetPhase,
+                            ArbitrationStrategy::RunPhaseImmediately);
+    }
+    return std::nullopt;
+  }
+
+  void pup(PUP::er& /*p*/) noexcept override {}
+};
+}  // namespace PhaseControl
+
+/// \cond
+template <typename Metavariables, typename Metavariables::Phase TargetPhase,
+          typename PhaseChangeRegistrars>
+PUP::able::PUP_ID PhaseControl::VisitAndReturn<
+    Metavariables, TargetPhase, PhaseChangeRegistrars>::my_PUP_ID = 0;
+/// \endcond
```

Comment:
Yeah, this was what I was getting at with the 'I'm not sure how to test' statement -- I think this really _ought_ to have a unit test, but because of the call to the main chare reduction in `contribute_phase_data_impl`, I'm not sure how to actually accomplish it.
As a half-measure I could implement a unit test that covers the other functions and include a comment that the contribution isn't tested but should be if we ever figure out a good way of mocking reductions.

Reply:
I think that would be good. Test what we easily can, deal with the rest when we figure out how :P